File: html/test_color_labeling.cpp
#include <iostream>
#include "camellia.h"

void cpp_example_color_labeling()
{
    CamImage source,YUV;
    CamRLEImage encoded;
    CamBlobs blobs;
    CamTable clusters;

    // Load picture alfa156.bmp
    source.load_bmp("resources/alfa156.bmp");
    source.to_yuv(YUV);

    // Set the color clusters
    const int limits[3*6]={
    //  Ymin Ymax Umin Umax Vmin Vmax
        0,   60,  0,   255, 0,   255, // Black
        230, 255, 0,   255, 0,   255, // White
        0,   255, 0,   255, 140, 255  // Red
    };
    clusters.set(limits,18);
    const int cluster_colors[3]={CAM_RGB(0,0,0),CAM_RGB(255,255,255),CAM_RGB(255,0,0)};
    
    // Threshold and encode
    encoded.encode_color(YUV,clusters);
    std::cout<<"Number of runs : "<<encoded.nbRuns<<std::endl;
    
    // Labeling
    encoded.labeling(blobs);
    std::cout<<blobs.nbBlobs<<" blobs detected"<<std::endl;

    // Print and draw the results 
    for (int i=0;i<blobs.nbBlobs;i++) {
        std::cout<<"Blob #"<<i<<" : Val="<<blobs.blobInfo[i].value<< \
            " ("<<blobs.blobInfo[i].left<<","<<blobs.blobInfo[i].top<<","<< \
            blobs.blobInfo[i].width<<","<<blobs.blobInfo[i].height<<") Surface="<<blobs.blobInfo[i].surface<< \
            std::endl;
        source.draw_rectangle(blobs.blobInfo[i].left, blobs.blobInfo[i].top,
            blobs.blobInfo[i].left+blobs.blobInfo[i].width-1,
            blobs.blobInfo[i].top+blobs.blobInfo[i].height-1,
            cluster_colors[blobs.blobInfo[i].value-1]);
    }

    // Save the result
    source.save_bmp("output/alfa156_color_labeling.bmp");
}

File: test/test_color_labeling.rb
require 'test/unit'
require 'lib/camellia'
include Camellia
class TestLabeling < Test::Unit::TestCase
  def test_labeling

    # load picture alfa156.bmp
    image=CamImage.new
    image.load_bmp("resources/alfa156.bmp")
    yuv=image.to_yuv
    
    # set color clusters
    clusters=CamTable.new
    clusters.set([
      # Ymin Ymax Umin Umax Vmin Vmax
        0,   60,  0,   255, 0,   255, # Black
        230, 255, 0,   255, 0,   255, # White
        0,   255, 0,   255, 140, 255  # Red
    ])
    cluster_colors=[cam_rgb(0,0,0),cam_rgb(255,255,255),cam_rgb(255,0,0)]
    
    # threshold and encode
    encoded=yuv.encode_color(clusters)
    puts "Number of runs : #{encoded.nb_runs}"
    
    # labeling
    blobs=encoded.labeling!    
    puts "#{blobs.nb_blobs} blobs detected"
    
    # print and deaw the results 
    i=0
    blobs.each {|b|
      puts "Blob #{i} : Val=#{b.value} (#{b.top},#{b.left},#{b.width},#{b.height}) Surface=#{b.surface}"
      image.draw_rectangle(b.left,b.top,b.left+b.width-1,b.top+b.height-1,cluster_colors[b.value-1])
      i=i+1
    }
    
    # save the result
    image.save_bmp("output/ruby_alfa156_labeling.bmp")
    
    # assertions
    assert_equal(encoded.nb_runs,3026)
    assert_equal(blobs.nb_blobs,190)
  end
end
